compiler/noirc_evaluator/src/ssa/opt/remove_bit_shifts.rs

```diff
@@ -148,7 +148,11 @@ impl Context<'_, '_, '_> {
             || {
                 // If we don't know `rhs`'s value then it could be anything up to the number
                 // of bits in the type, e.g. u32 means shifting by up to 32 bits as otherwise we get overflow.
-                self.context.dfg.type_of_value(rhs).bit_size()
+                let type_max = self.context.dfg.type_of_value(rhs).bit_size();
+                // get_value_max_num_bits might indicate that the underlying type actually has less
+                // bits than the RHS; e.g. because it was upcast from a u1 to a u32.
+                let value_max = 2u32.pow(self.context.dfg.get_value_max_num_bits(rhs));
```

Comment:
Would this overflow for u128, etc.?

Reply:
I only checked `u32` because that's what the result needs to be, and e.g. `2u32^32` does overflow, so it's only worth trying if the value is not greater than `type_max.ilog2()`, but since we know that it will only ever beat it with `u1`, I scrapped this idea.

Reply:
Aaaah... I started looking commit by commit, had to take my kid to school so I didn't see there were more commits. This looks good 👌